From e26aa2b60b58bdb4a9f2259fb313a5cfd6a65e42 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Fri, 6 Mar 2026 17:02:29 -0600
Subject: [PATCH 096/148] winewayland: Rework ensure contents to avoid
 compositor bugs.

Not a fan of having to change this but hyprland is hyprland i guess
---
 dlls/winewayland.drv/wayland_surface.c | 30 +++++++++++---------------
 1 file changed, 13 insertions(+), 17 deletions(-)

diff --git a/dlls/winewayland.drv/wayland_surface.c b/dlls/winewayland.drv/wayland_surface.c
index 8b2f51ac7f2..a422cae99a9 100644
--- a/dlls/winewayland.drv/wayland_surface.c
+++ b/dlls/winewayland.drv/wayland_surface.c
@@ -1496,7 +1496,7 @@ static const struct wl_buffer_listener dummy_buffer_listener =
 void wayland_surface_ensure_contents(struct wayland_surface *surface)
 {
     struct wayland_shm_buffer *dummy_shm_buffer;
-    HRGN damage;
+    HRGN damage = NULL;
     int width, height;
     BOOL needs_contents;
 
@@ -1511,28 +1511,24 @@ void wayland_surface_ensure_contents(struct wayland_surface *surface)
 
     if (!needs_contents) return;
 
-    /* Create a transparent dummy buffer. */
-    dummy_shm_buffer = wayland_shm_buffer_create(width, height, WL_SHM_FORMAT_ARGB8888);
-    if (!dummy_shm_buffer)
+    if (wayland_surface_reconfigure(surface))
     {
-        ERR("Failed to create dummy buffer\n");
-        return;
-    }
-    wl_buffer_add_listener(dummy_shm_buffer->wl_buffer, &dummy_buffer_listener,
-                           dummy_shm_buffer);
+        /* Create a transparent dummy buffer. */
+        dummy_shm_buffer = wayland_shm_buffer_create(width, height, WL_SHM_FORMAT_ARGB8888);
+        if (!dummy_shm_buffer)
+        {
+            ERR("Failed to create dummy buffer\n");
+            return;
+        }
+        wl_buffer_add_listener(dummy_shm_buffer->wl_buffer, &dummy_buffer_listener,
+                               dummy_shm_buffer);
 
-    if (!(damage = NtGdiCreateRectRgn(0, 0, width, height)))
-        WARN("Failed to create damage region for dummy buffer\n");
+        if (!(damage = NtGdiCreateRectRgn(0, 0, width, height)))
+            WARN("Failed to create damage region for dummy buffer\n");
 
-    if (wayland_surface_reconfigure(surface))
-    {
         wayland_surface_attach_shm(surface, dummy_shm_buffer, damage);
         wl_surface_commit(surface->wl_surface);
     }
-    else
-    {
-        wayland_shm_buffer_unref(dummy_shm_buffer);
-    }
 
     if (damage) NtGdiDeleteObjectApp(damage);
 }
-- 
2.54.0

